File: crates/yubikey-mgmt/src/device.rs
// Copyright (c) 2026 Yubico AB
// All rights reserved.
//
//   Redistribution and use in source and binary forms, with or
//   without modification, are permitted provided that the following
//   conditions are met:
//
//    1. Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//    2. Redistributions in binary form must reproduce the above
//       copyright notice, this list of conditions and the following
//       disclaimer in the documentation and/or other materials provided
//       with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

//! High-level device enumeration for YubiKeys.
//!
//! This module provides a convenient API for discovering connected YubiKeys
//! and opening sessions with them.
//!
//! # Example
//!
//! ```no_run
//! use yubikey_mgmt::device::list_devices;
//!
//! let devices = list_devices().unwrap();
//! for dev in &devices {
//!     println!("{} (serial: {:?})", dev.name(), dev.serial());
//! }
//! ```

use std::collections::HashSet;
use std::fmt;

use crate::iso7816::{SmartCardError, Transport, Version};
use crate::management::{Capability, DeviceInfo, FormFactor, ManagementSession};
use crate::transport::hid::{HidConnection, HidDeviceInfo, HidError, list_otp_devices};
use crate::transport::pcsc::{PcscConnection, PcscError, list_readers};

// ---------------------------------------------------------------------------
// DeviceError
// ---------------------------------------------------------------------------

/// Errors that can occur during device enumeration or connection.
#[derive(Debug)]
pub enum DeviceError {
    /// A SmartCard protocol error.
    SmartCard(SmartCardError),
    /// A PC/SC transport error.
    Pcsc(PcscError),
    /// A HID transport error.
    Hid(HidError),
    /// No YubiKey device was found.
    NoDeviceFound,
}

impl fmt::Display for DeviceError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::SmartCard(e) => write!(f, "SmartCard error: {e}"),
            Self::Pcsc(e) => write!(f, "PC/SC error: {e}"),
            Self::Hid(e) => write!(f, "HID error: {e}"),
            Self::NoDeviceFound => write!(f, "No YubiKey device found"),
        }
    }
}

impl std::error::Error for DeviceError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::SmartCard(e) => Some(e),
            Self::Pcsc(e) => Some(e),
            Self::Hid(e) => Some(e),
            Self::NoDeviceFound => None,
        }
    }
}

impl From<SmartCardError> for DeviceError {
    fn from(e: SmartCardError) -> Self {
        Self::SmartCard(e)
    }
}

impl From<PcscError> for DeviceError {
    fn from(e: PcscError) -> Self {
        Self::Pcsc(e)
    }
}

impl From<HidError> for DeviceError {
    fn from(e: HidError) -> Self {
        Self::Hid(e)
    }
}

// ---------------------------------------------------------------------------
// YubiKeyDevice
// ---------------------------------------------------------------------------

/// A discovered YubiKey that can open connections.
///
/// Represents a physical YubiKey found via PC/SC (CCID) and/or HID enumeration.
/// Use [`list_devices`] to discover connected devices.
#[derive(Debug, Clone)]
pub struct YubiKeyDevice {
    reader_name: Option<String>,
    hid_path: Option<String>,
    info: DeviceInfo,
}

impl YubiKeyDevice {
    /// Returns the [`DeviceInfo`] for this device.
    pub fn info(&self) -> &DeviceInfo {
        &self.info
    }

    /// Returns the serial number, if available.
    pub fn serial(&self) -> Option<u32> {
        self.info.serial
    }

    /// Returns the firmware version.
    pub fn version(&self) -> Version {
        self.info.version
    }

    /// Returns the product name derived from device info.
    pub fn name(&self) -> String {
        get_name(&self.info)
    }

    /// Returns the PC/SC reader name, if this device was found over CCID.
    pub fn reader_name(&self) -> Option<&str> {
        self.reader_name.as_deref()
    }

    /// Returns the HID device path, if this device was found over HID.
    pub fn hid_path(&self) -> Option<&str> {
        self.hid_path.as_deref()
    }

    /// Open a SmartCard (PC/SC) connection to this device.
    pub fn open_smartcard(&self) -> Result<PcscConnection, DeviceError> {
        let reader = self
            .reader_name
            .as_deref()
            .ok_or(DeviceError::NoDeviceFound)?;
        Ok(PcscConnection::new(reader, false)?)
    }

    /// Open an OTP HID connection to this device.
    pub fn open_otp(&self) -> Result<HidConnection, DeviceError> {
        let path = self
            .hid_path
            .as_deref()
            .ok_or(DeviceError::NoDeviceFound)?;
        Ok(HidConnection::new(path)?)
    }
}

impl fmt::Display for YubiKeyDevice {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name())?;
        if let Some(serial) = self.serial() {
            write!(f, " (serial: {serial})")?;
        }
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// Device enumeration
// ---------------------------------------------------------------------------

/// Discover all connected YubiKeys.
///
/// Scans PC/SC readers and HID devices, reads [`DeviceInfo`] from each,
/// and returns a unified list of discovered YubiKeys. Devices visible over
/// both PC/SC and HID (matched by serial number) are merged into a single
/// entry.
pub fn list_devices() -> Result<Vec<YubiKeyDevice>, DeviceError> {
    let mut devices = Vec::new();
    let mut seen_serials = HashSet::new();

    // Scan PC/SC readers
    if let Ok(readers) = list_readers() {
        for reader in readers {
            if !reader.to_ascii_lowercase().contains("yubi") {
                continue;
            }
            {
                if let Ok(info) = read_info(&reader) {
                    if let Some(serial) = info.serial {
                        seen_serials.insert(serial);
                    }
                    devices.push(YubiKeyDevice {
                        reader_name: Some(reader),
                        hid_path: None,
                        info,
                    });
                }
            }
        }
    }

    // Scan HID OTP devices, merging with existing PC/SC entries
    if let Ok(hid_devices) = list_otp_devices() {
        for hid in hid_devices {
            // If there is exactly one CCID device without a HID path, merge them.
            let should_merge = devices.len() == 1 && devices[0].hid_path.is_none();
            if should_merge {
                devices[0].hid_path = Some(hid.path);
            } else {
                let info = synthetic_hid_info(&hid);
                devices.push(YubiKeyDevice {
                    reader_name: None,
                    hid_path: Some(hid.path),
                    info,
                });
            }
        }
    }

    Ok(devices)
}

/// Build a minimal synthetic [`DeviceInfo`] for an HID-only device.
fn synthetic_hid_info(hid: &HidDeviceInfo) -> DeviceInfo {
    use std::collections::HashMap;

    let version = pid_to_version(hid.pid);
    let mut supported = HashMap::new();
    supported.insert(Transport::Usb, Capability::OTP);

    DeviceInfo {
        config: crate::management::DeviceConfig {
            enabled_capabilities: HashMap::new(),
            auto_eject_timeout: None,
            challenge_response_timeout: None,
            device_flags: None,
            nfc_restricted: None,
        },
        serial: None,
        version,
        form_factor: FormFactor::Unknown,
        supported_capabilities: supported,
        is_locked: false,
        is_fips: false,
        is_sky: false,
        part_number: None,
        fips_capable: Capability::NONE,
        fips_approved: Capability::NONE,
        pin_complexity: false,
        reset_blocked: Capability::NONE,
        fps_version: None,
        stm_version: None,
        version_qualifier: crate::management::VersionQualifier::final_release(version),
    }
}

/// Best-effort version guess from USB PID.
fn pid_to_version(pid: u16) -> Version {
    match pid {
        0x0110..=0x0113 => Version(3, 0, 0), // NEO family
        0x0116..=0x0120 => Version(3, 0, 0),
        0x0401..=0x0405 => Version(4, 0, 0), // YK4 family
        0x0406..=0x0410 => Version(5, 0, 0), // YK5 family
        _ => Version(0, 0, 0),
    }
}

// ---------------------------------------------------------------------------
// read_info
// ---------------------------------------------------------------------------

/// Read device info from a PC/SC reader.
///
/// Opens a fresh [`PcscConnection`] and uses [`ManagementSession`] to read
/// [`DeviceInfo`]. This works for YubiKey 4.1+ devices. For older devices a
/// fallback that scans individual applets would be needed (not yet implemented).
pub fn read_info(reader_name: &str) -> Result<DeviceInfo, DeviceError> {
    let conn = PcscConnection::new(reader_name, false)?;
    let mut session = ManagementSession::new(conn)?;
    // Use unchecked variant — dev devices report version 0.0.1 but still
    // support the DeviceInfo protocol.
    let info = session.read_device_info_unchecked()?;
    Ok(info)
}

// ---------------------------------------------------------------------------
// Product name logic
// ---------------------------------------------------------------------------

/// Preview firmware version ranges.
const PREVIEW_RANGES: &[(Version, Version)] = &[
    (Version(5, 0, 0), Version(5, 1, 0)),
    (Version(5, 2, 0), Version(5, 2, 3)),
    (Version(5, 5, 0), Version(5, 5, 2)),
];

fn is_preview(version: Version) -> bool {
    PREVIEW_RANGES
        .iter()
        .any(|(start, end)| version >= *start && version < *end)
}

fn fido_only(cap: Capability) -> bool {
    let non_fido = Capability::OTP.0
        | Capability::OATH.0
        | Capability::PIV.0
        | Capability::OPENPGP.0
        | Capability::HSMAUTH.0;
    let fido = Capability::U2F.0 | Capability::FIDO2.0;
    // No non-FIDO capabilities, but at least one FIDO capability
    (cap.0 & non_fido == 0) && (cap.0 & fido != 0)
}

/// Determine the product name of a YubiKey from its [`DeviceInfo`].
///
/// Ports the naming logic from the Python `yubikit.support.get_name`.
pub fn get_name(info: &DeviceInfo) -> String {
    let usb_supported = info
        .supported_capabilities
        .get(&Transport::Usb)
        .copied()
        .unwrap_or(Capability::NONE);

    let major = info.version.0;

    // Pre-YK4 devices
    if major < 4 {
        return if major == 0 {
            format!("YubiKey ({})", info.version)
        } else if major == 3 {
            "YubiKey NEO".to_string()
        } else {
            "YubiKey".to_string()
        };
    }

    // YK4 era
    if major == 4 {
        if info.is_fips {
            return "YubiKey FIPS (4 Series)".to_string();
        }
        if usb_supported == Capability(Capability::OTP.0 | Capability::U2F.0) {
            return "YubiKey Edge".to_string();
        }
        return "YubiKey 4".to_string();
    }

    // Preview firmware
    if is_preview(info.version) {
        return "YubiKey Preview".to_string();
    }

    // SKY without FIDO2 (SKY 1)
    if info.is_sky && !usb_supported.contains(Capability::FIDO2) {
        return "FIDO U2F Security Key".to_string();
    }

    // YK5+ dynamic naming (5.1.0+)
    if info.version >= Version(5, 1, 0) {
        return build_yk5_name(info, usb_supported);
    }

    // Fallback for 5.0.x
    if info.is_sky {
        "Security Key".to_string()
    } else {
        "YubiKey 5".to_string()
    }
}

fn build_yk5_name(info: &DeviceInfo, usb_supported: Capability) -> String {
    let is_nano = matches!(
        info.form_factor,
        FormFactor::UsbANano | FormFactor::UsbCNano
    );
    let is_bio = info.form_factor.is_bio();
    let is_c = matches!(
        info.form_factor,
        FormFactor::UsbCKeychain | FormFactor::UsbCNano | FormFactor::UsbCBio
    );
    let has_nfc = info.supported_capabilities.contains_key(&Transport::Nfc);

    let mut parts: Vec<&str> = Vec::new();

    // Base name
    if info.is_sky {
        parts.push("Security Key");
    } else {
        parts.push("YubiKey");
        if !is_bio {
            parts.push("5");
        }
    }

    // Connector type
    if is_c {
        parts.push("C");
    } else if info.form_factor == FormFactor::UsbCLightning {
        parts.push("Ci");
    }

    // Form factor / transport suffix
    if is_nano {
        parts.push("Nano");
    } else if has_nfc {
        parts.push("NFC");
    } else if info.form_factor == FormFactor::UsbAKeychain {
        parts.push("A");
    } else if is_bio {
        parts.push("Bio");
    }

    // Edition suffix
    if info.is_fips {
        parts.push("FIPS");
    } else if is_bio {
        if fido_only(usb_supported) {
            parts.push("- FIDO Edition");
        } else if usb_supported.contains(Capability::PIV) {
            parts.push("- Multi-protocol Edition");
        }
    } else if info.is_sky && info.serial.is_some() {
        parts.push("- Enterprise Edition");
    } else if info.pin_complexity && !info.is_sky {
        parts.push("- Enhanced PIN");
    }

    parts
        .join(" ")
        .replace("5 C", "5C")
        .replace("5 A", "5A")
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    fn make_info(
        version: Version,
        form_factor: FormFactor,
        is_sky: bool,
        is_fips: bool,
        serial: Option<u32>,
        nfc: bool,
        usb_cap: Capability,
        pin_complexity: bool,
    ) -> DeviceInfo {
        let mut supported = HashMap::new();
        supported.insert(Transport::Usb, usb_cap);
        if nfc {
            supported.insert(Transport::Nfc, usb_cap);
        }
        DeviceInfo {
            config: crate::management::DeviceConfig {
                enabled_capabilities: HashMap::new(),
                auto_eject_timeout: None,
                challenge_response_timeout: None,
                device_flags: None,
                nfc_restricted: None,
            },
            serial,
            version,
            form_factor,
            supported_capabilities: supported,
            is_locked: false,
            is_fips,
            is_sky,
            part_number: None,
            fips_capable: Capability::NONE,
            fips_approved: Capability::NONE,
            pin_complexity,
            reset_blocked: Capability::NONE,
            fps_version: None,
            stm_version: None,
            version_qualifier: crate::management::VersionQualifier::final_release(version),
        }
    }

    #[test]
    fn test_neo() {
        let info = make_info(
            Version(3, 5, 0),
            FormFactor::Unknown,
            false, false, Some(123), true,
            Capability(Capability::OTP.0 | Capability::OATH.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey NEO");
    }

    #[test]
    fn test_yk4() {
        let info = make_info(
            Version(4, 3, 7),
            FormFactor::UsbAKeychain,
            false, false, Some(456), false,
            Capability(Capability::OTP.0 | Capability::PIV.0 | Capability::OATH.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey 4");
    }

    #[test]
    fn test_yk4_fips() {
        let info = make_info(
            Version(4, 4, 5),
            FormFactor::UsbAKeychain,
            false, true, Some(789), false,
            Capability(Capability::OTP.0 | Capability::PIV.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey FIPS (4 Series)");
    }

    #[test]
    fn test_yk5_nfc() {
        let info = make_info(
            Version(5, 2, 4),
            FormFactor::UsbAKeychain,
            false, false, Some(100), true,
            Capability(Capability::OTP.0 | Capability::PIV.0 | Capability::FIDO2.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey 5 NFC");
    }

    #[test]
    fn test_yk5c_nano() {
        let info = make_info(
            Version(5, 4, 3),
            FormFactor::UsbCNano,
            false, false, Some(200), false,
            Capability(Capability::OTP.0 | Capability::PIV.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey 5C Nano");
    }

    #[test]
    fn test_yk5ci() {
        let info = make_info(
            Version(5, 2, 4),
            FormFactor::UsbCLightning,
            false, false, Some(300), false,
            Capability(Capability::OTP.0 | Capability::PIV.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey 5Ci");
    }

    #[test]
    fn test_security_key_nfc() {
        let info = make_info(
            Version(5, 2, 8),
            FormFactor::UsbAKeychain,
            true, false, None, true,
            Capability(Capability::U2F.0 | Capability::FIDO2.0),
            false,
        );
        assert_eq!(get_name(&info), "Security Key NFC");
    }

    #[test]
    fn test_bio_fido() {
        let info = make_info(
            Version(5, 5, 6),
            FormFactor::UsbABio,
            false, false, Some(400), false,
            Capability(Capability::U2F.0 | Capability::FIDO2.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey Bio - FIDO Edition");
    }

    #[test]
    fn test_bio_multi_protocol() {
        let info = make_info(
            Version(5, 6, 0),
            FormFactor::UsbCBio,
            false, false, Some(500), false,
            Capability(Capability::PIV.0 | Capability::FIDO2.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey C Bio - Multi-protocol Edition");
    }

    #[test]
    fn test_preview() {
        let info = make_info(
            Version(5, 0, 1),
            FormFactor::UsbAKeychain,
            false, false, Some(600), false,
            Capability(Capability::OTP.0 | Capability::PIV.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey Preview");
    }

    #[test]
    fn test_sky_enterprise() {
        let info = make_info(
            Version(5, 4, 3),
            FormFactor::UsbAKeychain,
            true, false, Some(700), false,
            Capability(Capability::U2F.0 | Capability::FIDO2.0),
            false,
        );
        assert_eq!(get_name(&info), "Security Key A - Enterprise Edition");
    }

    #[test]
    fn test_yk5a() {
        let info = make_info(
            Version(5, 2, 4),
            FormFactor::UsbAKeychain,
            false, false, Some(800), false,
            Capability(Capability::OTP.0 | Capability::PIV.0),
            false,
        );
        assert_eq!(get_name(&info), "YubiKey 5A");
    }

    #[test]
    fn test_fido_only() {
        assert!(fido_only(Capability(Capability::U2F.0 | Capability::FIDO2.0)));
        assert!(fido_only(Capability::FIDO2));
        assert!(!fido_only(Capability(Capability::FIDO2.0 | Capability::PIV.0)));
        assert!(!fido_only(Capability::NONE));
    }

    #[test]
    fn test_is_preview() {
        assert!(is_preview(Version(5, 0, 0)));
        assert!(is_preview(Version(5, 0, 1)));
        assert!(!is_preview(Version(5, 1, 0)));
        assert!(is_preview(Version(5, 2, 0)));
        assert!(is_preview(Version(5, 2, 2)));
        assert!(!is_preview(Version(5, 2, 3)));
        assert!(is_preview(Version(5, 5, 0)));
        assert!(is_preview(Version(5, 5, 1)));
        assert!(!is_preview(Version(5, 5, 2)));
        assert!(!is_preview(Version(5, 4, 0)));
    }

    #[test]
    fn test_enhanced_pin() {
        let info = make_info(
            Version(5, 7, 0),
            FormFactor::UsbCKeychain,
            false, false, Some(900), false,
            Capability(Capability::OTP.0 | Capability::PIV.0),
            true,
        );
        assert_eq!(get_name(&info), "YubiKey 5C - Enhanced PIN");
    }
}

File: packages/yubikit/yubikit/device.py
# Copyright (c) 2015-2020 Yubico AB
# All rights reserved.
#
#   Redistribution and use in source and binary forms, with or
#   without modification, are permitted provided that the following
#   conditions are met:
#
#    1. Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#    2. Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

"""
Device enumeration and native connection implementations for YubiKeys.
"""

from __future__ import annotations

import logging
from threading import Event
from typing import Callable, Iterable, Iterator, Mapping, TypeAlias

from _yubikit_native.device import NativeYubiKeyDevice as _NativeYubiKeyDeviceInfo
from _yubikit_native.device import list_devices as _native_list_devices
from _yubikit_native.device import scan_devices as _native_scan_devices
from _yubikit_native.hid import FidoConnection as _NativeFidoHidConnection
from _yubikit_native.hid import OtpConnection as _NativeOtpConnectionImpl
from _yubikit_native.pcsc import PcscConnection
from _yubikit_native.pcsc import list_readers as _native_list_readers
from yubikit.core import (
    PID,
    REINSERT_STATUS,
    TRANSPORT,
    CancelledException,
    Connection,
    YubiKeyDevice,
)
from yubikit.core.fido import FidoConnection, SmartCardCtapDevice
from yubikit.core.otp import OtpConnection
from yubikit.core.smartcard import SmartCardConnection
from yubikit.logging import LOG_LEVEL
from yubikit.management import DeviceInfo
from yubikit.support import read_info  # noqa: F401 - re-exported

logger = logging.getLogger(__name__)


# --- Connection implementations ---


class _NativeFidoConnection(FidoConnection):
    """FIDO connection backed by the native Rust CTAP HID transport."""

    def __init__(self, path: str, pid: int):
        self._native = _NativeFidoHidConnection(path, pid)
        self._device_version = self._native.device_version
        self._capabilities = self._native.capabilities
        self._path = path

    @property
    def capabilities(self) -> int:
        return self._capabilities

    @property
    def device_version(self) -> tuple[int, int, int]:
        return self._device_version

    def call(
        self,
        cmd: int,
        data: bytes = b"",
        event: Event | None = None,
        on_keepalive: Callable[[int], None] | None = None,
    ) -> bytes:
        return bytes(self._native.call(cmd, data))

    def close(self) -> None:
        self._native.close()

    @classmethod
    def list_devices(cls) -> Iterator[_NativeFidoConnection]:
        from _yubikit_native.hid import list_fido_devices as _native_list_fido_devices

        for dev in _native_list_fido_devices():
            yield cls(dev.path, dev.pid)


class _NativeOtpConnection(OtpConnection):
    """OTP connection backed by the Rust HID implementation."""

    def __init__(self, path: str):
        self._path = path
        self._native = _NativeOtpConnectionImpl(path)

    def close(self) -> None:
        self._native.close()

    def receive(self) -> bytes:
        data = bytes(self._native.get_feature_report())
        logger.log(LOG_LEVEL.TRAFFIC, "RECV: %s", data.hex())
        return data

    def send(self, data: bytes) -> None:
        logger.log(LOG_LEVEL.TRAFFIC, "SEND: %s", data.hex())
        self._native.set_feature_report(data)


YK_READER_NAME = "yubico yubikey"


class _NativeSmartCardConnection(SmartCardConnection):
    def __init__(self, reader_name):
        self._native = PcscConnection.open(reader_name)
        self._transport = (
            TRANSPORT.USB if self._native.transport == "usb" else TRANSPORT.NFC
        )

    @property
    def transport(self):
        return self._transport

    def close(self):
        self._native.disconnect()

    def send_and_receive(self, apdu):
        """Sends a command APDU and returns the response data and sw"""
        logger.log(LOG_LEVEL.TRAFFIC, "SEND: %s", apdu.hex())
        resp = self._native.transmit(apdu)
        data = resp[:-2]
        sw = resp[-2] << 8 | resp[-1]
        logger.log(LOG_LEVEL.TRAFFIC, "RECV: %s SW=%04x", data.hex(), sw)
        return bytes(data), sw


def list_readers():
    try:
        return _native_list_readers()
    except OSError:
        return []


# --- Device enumeration ---


_T_CONNECTION: TypeAlias = type[Connection] | type[FidoConnection]

# Map connection types to native transport names for list_devices
_CONNECTION_TRANSPORT_MAP: dict[_T_CONNECTION, str] = {
    SmartCardConnection: "ccid",
    OtpConnection: "otp",
    FidoConnection: "fido",
}

_DEFAULT_CONNECTION_TYPES: list[_T_CONNECTION] = list(_CONNECTION_TRANSPORT_MAP.keys())


def scan_devices() -> tuple[Mapping[PID, int], int]:
    """Scan USB for attached YubiKeys, without opening any connections.

    The native implementation handles Windows non-admin FIDO fallback.

    :return: A dict mapping PID to device count, and a state object which can be
        used to detect changes in attached devices.
    """
    raw_counts, state = _native_scan_devices()

    # Convert raw PID ints to PID enum values
    merged: dict[PID, int] = {}
    for pid_int, count in raw_counts.items():
        try:
            merged[PID(pid_int)] = count
        except ValueError:
            logger.debug(f"Unsupported PID: {pid_int:#04x}")

    return merged, state


class _NativeYubiKeyDevice(YubiKeyDevice):
    """YubiKey device backed by native Rust enumeration."""

    def __init__(self, native_dev: _NativeYubiKeyDeviceInfo, info: DeviceInfo):
        self._pid = PID(native_dev.pid) if native_dev.pid else None
        self._fingerprint = (
            native_dev.reader_name or native_dev.hid_path or native_dev.fido_path or ""
        )
        self._transport = (
            TRANSPORT.NFC if native_dev.transport == "nfc" else TRANSPORT.USB
        )
        self._native = native_dev
        self._info = info

    @property
    def transport(self) -> TRANSPORT:
        return self._transport

    @property
    def pid(self) -> PID | None:
        return self._pid

    @property
    def fingerprint(self) -> str:
        return self._fingerprint

    @property
    def reader_name(self) -> str | None:
        return self._native.reader_name

    def supports_connection(self, connection_type: type) -> bool:
        if issubclass(connection_type, SmartCardConnection):
            return self._native.reader_name is not None
        if issubclass(connection_type, OtpConnection):
            return self._native.hid_path is not None
        if issubclass(connection_type, FidoConnection):
            return self._native.fido_path is not None
        return False

    def open_connection(self, connection_type):  # type: ignore[override]
        assert isinstance(connection_type, type)  # noqa: S101
        if issubclass(connection_type, SmartCardConnection):
            reader = self._native.reader_name
            if reader is not None:
                return _NativeSmartCardConnection(reader)
        if issubclass(connection_type, FidoConnection):
            fido_path = self._native.fido_path
            if fido_path is not None and self.pid is not None:
                if issubclass(connection_type, SmartCardCtapDevice):
                    reader = self._native.reader_name
                    if reader is not None:
                        return SmartCardCtapDevice(_NativeSmartCardConnection(reader))
                return _NativeFidoConnection(fido_path, self.pid)
        if issubclass(connection_type, OtpConnection):
            hid_path = self._native.hid_path
            if hid_path is not None:
                return _NativeOtpConnection(hid_path)
        raise ValueError(f"Unsupported connection type: {connection_type}")

    def reinsert(
        self,
        reinsert_cb: Callable[[REINSERT_STATUS], None] | None = None,
        event: Event | None = None,
    ) -> None:
        cb = reinsert_cb or (lambda _: None)
        ev = event or Event()
        status_map = {
            "remove": REINSERT_STATUS.REMOVE,
            "reinsert": REINSERT_STATUS.REINSERT,
        }
        try:
            self._native.reinsert(
                lambda s: cb(status_map[s]),
                lambda: ev.is_set(),
            )
        except RuntimeError as e:
            msg = str(e)
            if "cancelled" in msg.lower():
                raise CancelledException() from e
            if "different" in msg.lower():
                raise ValueError(msg) from e
            raise


def _device_info_from_native_dev(native_dev: _NativeYubiKeyDeviceInfo) -> DeviceInfo:
    """Convert a native device's info dict to a DeviceInfo."""
    from yubikit.management import _device_info_from_native

    return _device_info_from_native(native_dev.info())


def list_all_devices(
    connection_types: Iterable[_T_CONNECTION] = _DEFAULT_CONNECTION_TYPES,
) -> list[tuple[YubiKeyDevice, DeviceInfo]]:
    """Connect to all attached YubiKeys and read device info from them.

    :param connection_types: An iterable of YubiKey connection types.
    :return: A list of (device, info) tuples for each connected device.
    """
    transports = [
        _CONNECTION_TRANSPORT_MAP[ct]
        for ct in connection_types
        if ct in _CONNECTION_TRANSPORT_MAP
    ]
    if not transports:
        return []

    results: list[tuple[YubiKeyDevice, DeviceInfo]] = []
    for native_dev in _native_list_devices(transports):
        info = _device_info_from_native_dev(native_dev)
        results.append((_NativeYubiKeyDevice(native_dev, info), info))
    return results
